server/aesdsocket.c

```diff
@@ -0,0 +1,317 @@
+    int file_size = 0;
+
+    // String to hold the IP address of the client; used when printing logs
+    char address_string[INET_ADDRSTRLEN];
+
+    while(1) {
+
+        // Continuously restarting connections in the while(1) loop
+        new_fd = accept(socketfd, (struct sockaddr *)&test_addr, &addr_size);
+
+        if (new_fd == -1) {
+            perror("accept");
+            continue;
+        } 
+
+        // Using sock_addr_in and inet_ntop function to print the IP address of the accepted client connection
+        // Reference: https://stackoverflow.com/questions/2104495/extract-ip-from-connection-that-listen-and-accept-in-socket-programming-in-linux
+        struct sockaddr_in *p = (struct sockaddr_in *)&test_addr;
+        syslog(LOG_DEBUG, "Accepted connection from %s", inet_ntop(AF_INET, &p->sin_addr, address_string, sizeof(address_string)));
+
+        // While loop until bytes received is <= 0
+        while ((bytes_received = recv(new_fd, buffer, sizeof(buffer), 0)) > 0 ) {
+
+            printf("Bytes Received = %d\n", bytes_received);
+
+            // Iterating over bytes received to check for "\n" character
+            for (int i = 0; i < bytes_received; i++) {
+                if (buffer[i] == '\n') {
+                    packet_complete_flag = 1;
+                    temp_iterator = i+1;
+                    break;
+                }
+            }
+
+            // Copying from receive buffer to storage buffer 
+            memcpy(storage_buffer + (storage_buffer_size_count - 1)*INITIAL_BUFFER_SIZE, buffer, bytes_received);
+
+            // Check if data packet is complete and assign bytes to write based on temporary iterator
+            if (packet_complete_flag == 1) {
+                bytes_to_write = (storage_buffer_size_count - 1)*INITIAL_BUFFER_SIZE + temp_iterator;
+                break;
+            } else {
+                // Increasing size of storage buffer by INITIAL_BUFFER_SIZE amount if no "\n" character found in buffer
+                char *temp_ptr  = (char *)realloc(storage_buffer, (storage_buffer_size + INITIAL_BUFFER_SIZE));
+                if (temp_ptr) {
+                    storage_buffer = temp_ptr;
+                    storage_buffer_size += INITIAL_BUFFER_SIZE;
+                    storage_buffer_size_count += 1;
+                } else {
+                    printf("Unable to allocate memory\n");
+                }
+            }
+        }
+
+        printf("Storage Buffer size = %d\n", bytes_to_write);
+
+        // Check if packet_complete; indicating "\n" encountered in received data
+        if (packet_complete_flag == 1) {
+
+            packet_complete_flag = 0;
+
+            //Writing to file
+            int nr = write(testfile_fd, storage_buffer, bytes_to_write);
+            file_size += nr; // Adding to current file length
+
+            lseek(testfile_fd, 0, SEEK_SET); // Setting the FD to start of file
+
+            char *read_buffer = (char *)malloc(file_size);
```

Comment:
@GautamaGandhi what assumption are you making here about the size of all packets?

Reply:
I am mallocing the amount based on bytes written in the write command. This is being incremented for every subsequent data packet. This could fail if there is not enough space in the heap. A loop with a fixed or small amount of malloced data should work.